cf-custom-resources/lib/cert-replicator.js

```diff
@@ -0,0 +1,320 @@
+  try {
+    let inUseByResources;
+
+    for (let attempt = 0; attempt < maxAttempts; attempt++) {
+      const { Certificate } = await acm
+        .describeCertificate({
+          CertificateArn: arn,
+        })
+        .promise();
+
+      inUseByResources = Certificate.InUseBy || [];
+      if (!inUseByResources.length) {
+        break;
+      }
+      // Deleting resources can be quite slow - so just sleep 30 seconds between checks.
+      await sleep(30000);
+    }
+    if (inUseByResources.length) {
+      throw new Error(
+        `Certificate still in use by ${inUseByResources.join()} after checking for ${maxAttempts} attempts.`
+      );
+    }
+
+    await acm
+      .deleteCertificate({
+        CertificateArn: arn,
+      })
+      .promise();
+  } catch (err) {
+    if (err.name !== "ResourceNotFoundException") {
+      throw err;
+    }
+  }
+};
+
+const validateCertificate = async function (certificateARN, acm) {
+  await acm
+    .waitFor("certificateValidated", {
+      // Wait up to 9 minutes and 30 seconds
+      $waiter: {
+        delay: 30,
+        maxAttempts: 19,
+      },
+      CertificateArn: certificateARN,
+    })
+    .promise();
+};
+
+/**
+ * Main certificate replicator handler, invoked by Lambda
+ */
+exports.certificateReplicateHandler = async function (event, context) {
+  let responseData = {};
+  let physicalResourceId = event.PhysicalResourceId;
+  const props = event.ResourceProperties;
+  const [targetRegion, envRegion, certArn] = [
+    props.TargetRegion,
+    props.EnvRegion,
+    props.CertificateArn,
+  ];
+  let handler = async function () {
+    // Configure clients.
+    const envRegionAcm = new aws.ACM({ envRegion });
+    const targetRegionAcm = new aws.ACM({ targetRegion });
+    switch (event.RequestType) {
+      case "Create":
+      case "Update":
+        const response = await replicateCertificate(
+          event.RequestId,
+          props.AppName,
+          props.EnvName,
+          certArn,
+          envRegionAcm,
+          targetRegionAcm
+        );
+        responseData.Arn = physicalResourceId = response.CertificateArn;
+        await validateCertificate(response.CertificateArn, targetRegionAcm);
+        break;
+      case "Delete":
+        // If the resource didn't create correctly, the physical resource ID won't be the
+        // certificate ARN, so don't try to delete it in that case.
+        if (physicalResourceId.startsWith("arn:")) {
+          await deleteCertificate(physicalResourceId, targetRegionAcm);
+        }
+        break;
+      default:
+        throw new Error(`Unsupported request type ${event.RequestType}`);
+    }
+  };
+  try {
+    await Promise.race([exports.deadlineExpired(), handler()]);
+    await report(event, context, "SUCCESS", physicalResourceId, responseData);
+  } catch (err) {
+    console.log(`Caught error ${err}.`);
+    await report(
+      event,
+      context,
+      "FAILED",
+      physicalResourceId,
```

Comment:
so if it's a `CREATE` event and  `replicateCertificate` fails then the `physicalResourceId` will be `undefined`. What happens then?
Do we need to keep:
```
let physicalResourceID = event.PhysicalResourceId || context.logStreamName; 
```

Reply:
I guess it'll be fine? Just checked the other lambda i'll keep it for consistency